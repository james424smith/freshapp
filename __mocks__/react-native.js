import * as ReactNative from "react-native";

export * from "react-native";
export const Vibration = { vibrate: jest.fn() };
export const Adnimated = { spring: jest.fn() };

export default Object.setPrototypeOf(
  {
    Vibration,
    Adnimated,
  },
  ReactNative
);
