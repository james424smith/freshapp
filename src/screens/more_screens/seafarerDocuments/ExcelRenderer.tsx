import React, { useState, useEffect } from "react";
import { View, Platform } from "react-native";
import { Document, SeafarerDocumentsFile } from "../../../interfaces";
import OpenFile from "@quinaryio/react-native-doc-viewer";
import LoadingScreen from "../../../components/LoadingScreen";
import styles from "../../../styles/more_screens/SalaryPdfStyles";
import { getSeafarerDocumentsFile } from "../../../redux/actions";
import { useSelector, useDispatch } from "react-redux";
import { IRootReducerType } from "../../../redux/reducers";

import { IOS } from "../../../constants/Platforms";
import Text from "../../../components/StyledText";
import { RouteProp } from "@react-navigation/native";
import { DocumentsStackParamList } from "../../../navigation/MoreScreenNavigation";
const logoSize = 100;

interface Props {
  route: RouteProp<DocumentsStackParamList, "SeafarerDocumentsExcelFile">;
}

const ExcelRendererComponent = (props: Props) => {
  const { route } = props;
  const seafarerDocumentsFile = useSelector<
    IRootReducerType,
    SeafarerDocumentsFile | undefined
  >(({ seafarerDocuments }) => seafarerDocuments.seafarerDocumentsFile);
  const isLoadingNew = useSelector<IRootReducerType, boolean>(
    ({ seafarerDocuments }) => seafarerDocuments.isLoadingNew
  );
  const dispatch = useDispatch();
  const [docDetails, setDocumentDetails] = useState<Document | undefined>(
    undefined
  );
  const [animating, setAnimating] = useState(false);
  const [error, setError] = useState<string>("");

  useEffect(() => {
    const documentDetails = route.params?.documentDetails;
    setDocumentDetails(documentDetails);
    dispatch(getSeafarerDocumentsFile(documentDetails, false));
  }, [route.params, dispatch]);

  useEffect(() => {
    const handleError = (errorLocal: any) => {
      if (errorLocal) {
        setAnimating(false);
        setError("error");
      } else {
        setAnimating(false);
      }
    };

    setAnimating(true);
    if (seafarerDocumentsFile && docDetails && !isLoadingNew) {
      if (Platform.OS === IOS) {
        OpenFile.openDocb64(
          [
            {
              base64: seafarerDocumentsFile.document,
              fileName: `doc-${docDetails.documentId}-counter-${
                docDetails.documentCounter
              }.${seafarerDocumentsFile.fileType.toLowerCase()}`,
              fileType: seafarerDocumentsFile.fileType.toLowerCase(),
            },
          ],
          (errorLocal: any, _url: any) => {
            handleError(errorLocal);
          }
        );
      } else {
        OpenFile.openDocb64(
          [
            {
              base64: seafarerDocumentsFile.document,
              fileName: `doc-${docDetails.documentId}-counter-${
                docDetails.documentCounter
              }.${seafarerDocumentsFile.fileType.toLowerCase()}`,
              fileType: seafarerDocumentsFile.fileType.toLowerCase(),
              // eslint-disable-next-line @typescript-eslint/ban-ts-comment
              // @ts-ignore
              cache: false,
            },
          ],
          (errorLocal: any, _url: any) => {
            handleError(errorLocal);
          }
        );
      }
    }
  }, [seafarerDocumentsFile, docDetails, isLoadingNew]);

  if (error) {
    return (
      <View style={styles.root}>
        <Text>
          Your phone cannot open {seafarerDocumentsFile?.fileType} files. Please
          download an app that is capable of opening such files and try again.
        </Text>
      </View>
    );
  } else if (animating || isLoadingNew) {
    return (
      <View style={styles.root}>
        <LoadingScreen logoWidthSize={logoSize} logoHeightSize={logoSize} />
      </View>
    );
  } else {
    return (
      <View style={styles.root}>
        <Text>
          {docDetails?.description} is open in an external{" "}
          {docDetails?.documentExtension?.toLowerCase()} app.
        </Text>
      </View>
    );
  }
};

export default ExcelRendererComponent;
