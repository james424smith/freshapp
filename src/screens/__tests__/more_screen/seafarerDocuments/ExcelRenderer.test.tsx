import React from "react";
import ExcelRenderer from "../../../more_screens/seafarerDocuments/ExcelRenderer";
import { render, cleanup, waitFor } from "@testing-library/react-native";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import { Platform } from "react-native";
import * as OpenFile from "@quinaryio/react-native-doc-viewer";

const document = {
  documentId: 1234,
  documentCounter: 12,
  docLicNumber: "1234",
  nation: "CY",
  description: "description",
  expiryDate: "01-01-2021",
  issueDate: "01-01-2019",
  docFileSize: "1mb",
  isDeactivated: false,
  categoryNumber: "4",
  fileType: "pdf",
  isExpired: false,
  isOffline: false,
  documentExtension: "pdf",
  marked: false,
  hasAttachment: true,
};

describe("ExcelRenderer snapshot test", () => {
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
    jest.restoreAllMocks();
  });

  it("should render correct ui for ExcelRenderer when document is present and downloaded and loading is false and platform is IOS", async () => {
    Platform.OS = "ios";
    jest.useRealTimers();
    const route = {
      key: "SeafarerDocumentsExcelFile-wPFrEJG5kjiOrE_Qk2kVs",
      name: "SeafarerDocumentsExcelFile" as "SeafarerDocumentsExcelFile",
      params: {
        documentDetails: { ...document, isOffline: true },
      },
    };

    const mockedState = {
      seafarerDocuments: {
        seafarerDocumentsFile: {
          document: "1234",
          fileType: "pdf",
        },
        isLoadingNew: false,
      },
      network: { isConnected: true },
    };
    const mockedStore = configureMockStore();

    const { toJSON } = render(
      <Provider store={mockedStore(mockedState)}>
        <ExcelRenderer route={route} />
      </Provider>
    );
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
  it("should render correct ui for ExcelRenderer when document is present and downloaded and loading is false and platform is IOS and OpenDoc throws an error", async () => {
    Platform.OS = "ios";
    jest.useRealTimers();
    const route = {
      key: "SeafarerDocumentsExcelFile-wPFrEJG5kjiOrE_Qk2kVs",
      name: "SeafarerDocumentsExcelFile" as "SeafarerDocumentsExcelFile",
      params: {
        documentDetails: { ...document, isOffline: true },
      },
    };

    const spy = jest
      .spyOn(OpenFile, "openDocb64")
      .mockImplementation((_a, e) => {
        return e("e", "");
      });

    const mockedState = {
      seafarerDocuments: {
        seafarerDocumentsFile: {
          document: "1234",
          fileType: "pdf",
        },
        isLoadingNew: false,
      },
      network: { isConnected: true },
    };
    const mockedStore = configureMockStore();

    const { toJSON } = render(
      <Provider store={mockedStore(mockedState)}>
        <ExcelRenderer route={route} />
      </Provider>
    );
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
    spy.mockRestore();
  });
  it("should render correct ui for ExcelRenderer when document is present and downloaded and loading is false and platform is android and OpenDoc throws an error", async () => {
    Platform.OS = "android";
    jest.useRealTimers();
    const route = {
      key: "SeafarerDocumentsExcelFile-wPFrEJG5kjiOrE_Qk2kVs",
      name: "SeafarerDocumentsExcelFile" as "SeafarerDocumentsExcelFile",
      params: {
        documentDetails: { ...document, isOffline: true },
      },
    };

    const spy = jest
      .spyOn(OpenFile, "openDocb64")
      .mockImplementation((_a, e) => {
        return e("", "");
      });
    const mockedState = {
      seafarerDocuments: {
        seafarerDocumentsFile: {
          document: "1234",
          fileType: "pdf",
        },
        isLoadingNew: false,
      },
      network: { isConnected: true },
    };
    const mockedStore = configureMockStore();

    const { toJSON } = render(
      <Provider store={mockedStore(mockedState)}>
        <ExcelRenderer route={route} />
      </Provider>
    );
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
    spy.mockRestore();
  });
  it("should render correct ui for ExcelRenderer when document is present and downloaded and loading is false and platform is android", async () => {
    Platform.OS = "android";
    jest.useRealTimers();
    const route = {
      key: "SeafarerDocumentsExcelFile-wPFrEJG5kjiOrE_Qk2kVs",
      name: "SeafarerDocumentsExcelFile" as "SeafarerDocumentsExcelFile",
      params: {
        documentDetails: { ...document, isOffline: true },
      },
    };

    const mockedState = {
      seafarerDocuments: {
        seafarerDocumentsFile: {
          document: "1234",
          fileType: "pdf",
        },
        isLoadingNew: false,
      },
      network: { isConnected: true },
    };
    const mockedStore = configureMockStore();

    const { toJSON } = render(
      <Provider store={mockedStore(mockedState)}>
        <ExcelRenderer route={route} />
      </Provider>
    );
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it("should render correct ui for ExcelRenderer when document is present and not downloaded and loading is false", () => {
    const route = {
      key: "SeafarerDocumentsExcelFile-wPFrEJG5kjiOrE_Qk2kVs",
      name: "SeafarerDocumentsExcelFile" as "SeafarerDocumentsExcelFile",
      params: {
        documentDetails: document,
      },
    };
    const mockedState = {
      seafarerDocuments: {
        seafarerDocumentsFile: {
          document: "1234",
          fileType: "pdf",
        },
        isLoadingNew: false,
      },
      network: { isConnected: true },
    };
    const mockedStore = configureMockStore();

    const { toJSON } = render(
      <Provider store={mockedStore(mockedState)}>
        <ExcelRenderer route={route} />
      </Provider>
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it("should render correct ui for ExcelRenderer when document is undefined and loading is false", () => {
    const route = {
      key: "SeafarerDocumentsExcelFile-wPFrEJG5kjiOrE_Qk2kVs",
      name: "SeafarerDocumentsExcelFile" as "SeafarerDocumentsExcelFile",
      params: {
        documentDetails: document,
      },
    };
    const mockedState = {
      seafarerDocuments: {
        seafarerDocumentsFile: undefined,
        isLoadingNew: false,
      },
      network: { isConnected: true },
    };
    const mockedStore = configureMockStore();

    const { toJSON } = render(
      <Provider store={mockedStore(mockedState)}>
        <ExcelRenderer route={route} />
      </Provider>
    );

    expect(toJSON()).toMatchSnapshot();
  });
  it("should render correct ui for ExcelRenderer when document is present and loading is true", () => {
    const route = {
      key: "SeafarerDocumentsExcelFile-wPFrEJG5kjiOrE_Qk2kVs",
      name: "SeafarerDocumentsExcelFile" as "SeafarerDocumentsExcelFile",
      params: {
        documentDetails: document,
      },
    };
    const mockedState = {
      seafarerDocuments: {
        seafarerDocumentsFile: { document: "1234", fileType: "pdf" },
        isLoadingNew: true,
      },
      network: { isConnected: true },
    };
    const mockedStore = configureMockStore();

    const { toJSON } = render(
      <Provider store={mockedStore(mockedState)}>
        <ExcelRenderer route={route} />
      </Provider>
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
