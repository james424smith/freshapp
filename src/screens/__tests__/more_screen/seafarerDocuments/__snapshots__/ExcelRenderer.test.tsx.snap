// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExcelRenderer snapshot test should render correct ui for ExcelRenderer when document is present and downloaded and loading is false and platform is IOS 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "margin": 10,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "marginBottom": 112.5,
      }
    }
  >
    <Image
      source={Object {}}
      style={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    />
  </View>
</View>
`;

exports[`ExcelRenderer snapshot test should render correct ui for ExcelRenderer when document is present and downloaded and loading is false and platform is IOS and OpenDoc throws an error 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "margin": 10,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#2A2E43",
          "fontFamily": "Roboto-Regular",
        },
        undefined,
      ]
    }
  >
    Your phone cannot open 
    pdf
     files. Please download an app that is capable of opening such files and try again.
  </Text>
</View>
`;

exports[`ExcelRenderer snapshot test should render correct ui for ExcelRenderer when document is present and downloaded and loading is false and platform is android 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "margin": 10,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "marginBottom": 112.5,
      }
    }
  >
    <Image
      source={Object {}}
      style={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    />
  </View>
</View>
`;

exports[`ExcelRenderer snapshot test should render correct ui for ExcelRenderer when document is present and downloaded and loading is false and platform is android and OpenDoc throws an error 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "margin": 10,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#2A2E43",
          "fontFamily": "Roboto-Regular",
        },
        undefined,
      ]
    }
  >
    description
     is open in an external
     
    pdf
     app.
  </Text>
</View>
`;

exports[`ExcelRenderer snapshot test should render correct ui for ExcelRenderer when document is present and loading is true 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "margin": 10,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "marginBottom": 112.5,
      }
    }
  >
    <Image
      source={Object {}}
      style={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    />
  </View>
</View>
`;

exports[`ExcelRenderer snapshot test should render correct ui for ExcelRenderer when document is present and not downloaded and loading is false 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "margin": 10,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "marginBottom": 112.5,
      }
    }
  >
    <Image
      source={Object {}}
      style={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    />
  </View>
</View>
`;

exports[`ExcelRenderer snapshot test should render correct ui for ExcelRenderer when document is undefined and loading is false 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "margin": 10,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "marginBottom": 112.5,
      }
    }
  >
    <Image
      source={Object {}}
      style={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    />
  </View>
</View>
`;
